trainer/lib/trainer/plist_test_summary_parser.rb

```diff
@@ -0,0 +1,84 @@
+require 'plist'
+
+module Trainer
+  module PlistTestSummaryParser
+    class << self
+      def parse_content(raw_json, xcpretty_naming)
+        data = raw_json["TestableSummaries"].collect do |testable_summary|
+          summary_row = {
+            project_path: testable_summary["ProjectPath"],
+            target_name: testable_summary["TargetName"],
+            test_name: testable_summary["TestName"],
+            duration: testable_summary["Tests"].map { |current_test| current_test["Duration"] }.inject(:+),
+            tests: unfold_tests(testable_summary["Tests"]).collect do |current_test|
+              test_group, test_name = test_group_and_name(testable_summary, current_test, xcpretty_naming)
+              current_row = {
+                identifier: current_test["TestIdentifier"],
+                test_group: test_group,
+                name: test_name,
+                object_class: current_test["TestObjectClass"],
+                status: current_test["TestStatus"],
+                guid: current_test["TestSummaryGUID"],
+                duration: current_test["Duration"]
+              }
+              if current_test["FailureSummaries"]
+                current_row[:failures] = current_test["FailureSummaries"].collect do |current_failure|
+                  {
+                    file_name: current_failure['FileName'],
+                    line_number: current_failure['LineNumber'],
+                    message: current_failure['Message'],
+                    performance_failure: current_failure['PerformanceFailure'],
+                    failure_message: "#{current_failure['Message']} (#{current_failure['FileName']}:#{current_failure['LineNumber']})"
+                  }
+                end
+              end
+              current_row
+            end
+          }
+          summary_row[:number_of_tests] = summary_row[:tests].count
+          summary_row[:number_of_failures] = summary_row[:tests].find_all { |a| (a[:failures] || []).count > 0 }.count
+
+          # Makes sure that plist support matches data output of xcresult
+          summary_row[:number_of_tests_excluding_retries] = summary_row[:number_of_tests]
+          summary_row[:number_of_failures_excluding_retries] = summary_row[:number_of_failures]
+          summary_row[:number_of_retries] = 0
+
+          summary_row
+        end
+        data
+      end
+
+      def ensure_file_valid!(raw_json)
+        format_version = raw_json["FormatVersion"]
+        supported_versions = ["1.1", "1.2"]
```

Comment:
Does it mean that with every new version of Xcode, we need to check if the format has changed and support the new version if necessary?

Reply:
Yes and no.

To the best of my understanding, this `.plist` format is legacy format, even older than before `.xcresult` bundles and `xcresulttool` were introduced. This is just existing code that was initially in `xcresult.rb` and that I extracted in its own file just like I did for `legacy_xcresult.rb` so we keep that backwards compatibility, but in practice I'm not sure any modern version of Xcode produces those `.plist` file formats for test summaries anymore, so I don't see a new `FormatVersion` being added ever by now.